pkg/controller/handler.go

```diff
@@ -4,72 +4,70 @@ import (
 	"github.com/containous/maesh/pkg/k8s"
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
 )
 
+// configMessageChanRebuild rebuild.
+const configMessageChanRebuild = "rebuild"
+
 // Handler is an implementation of a ResourceEventHandler.
 type Handler struct {
 	log               logrus.FieldLogger
-	ignored           k8s.IgnoreWrapper
+	ignoredResources  k8s.IgnoreWrapper
 	configRefreshChan chan string
 	serviceManager    ServiceManager
 }
 
 // NewHandler creates a handler.
 func NewHandler(log logrus.FieldLogger, ignored k8s.IgnoreWrapper, serviceManager ServiceManager, configRefreshChan chan string) *Handler {
-	h := &Handler{
+	return &Handler{
 		log:               log,
-		ignored:           ignored,
+		ignoredResources:  ignored,
 		configRefreshChan: configRefreshChan,
 		serviceManager:    serviceManager,
 	}
-
-	if err := h.Init(); err != nil {
-		log.Errorln("Could not initialize MeshControllerHandler")
-	}
-
-	return h
 }
 
-// Init handles any handler initialization.
-func (h *Handler) Init() error {
-	h.log.Debugln("MeshControllerHandler.Init")
-
-	return nil
-}
-
-// OnAdd executed when an object is added.
+// OnAdd is called when an object is added.
 func (h *Handler) OnAdd(obj interface{}) {
-	// assert the type to an object to pull out relevant data
-	switch obj := obj.(type) {
-	case *corev1.Service:
-		if h.ignored.IsIgnored(obj.ObjectMeta) {
-			return
-		}
+	if h.isIgnoredResource(obj) {
+		return
+	}
 
+	// If the created object is a service we should create a corresponding shadow service.
+	if obj, isService := obj.(*corev1.Service); isService {
 		if err := h.serviceManager.Create(obj); err != nil {
 			h.log.Errorf("Could not create mesh service: %v", err)
 		}
-	case *corev1.Endpoints:
-		return
-	case *corev1.Pod:
-		if !isMeshPod(obj) {
-			return
-		}
 	}
 
 	// Trigger a configuration rebuild.
-	h.configRefreshChan <- k8s.ConfigMessageChanRebuild
+	h.configRefreshChan <- configMessageChanRebuild
 }
 
-// OnUpdate executed when an object is updated.
+// OnUpdate is called when an object is updated and ensures that the proper handler is called depending on the filter matches.
 func (h *Handler) OnUpdate(oldObj, newObj interface{}) {
-	// Assert the type to an object to pull out relevant data.
-	switch obj := newObj.(type) {
-	case *corev1.Service:
-		if h.ignored.IsIgnored(obj.ObjectMeta) {
-			return
-		}
+	isOldIgnored := h.isIgnoredResource(oldObj)
+	isNewIgnored := h.isIgnoredResource(newObj)
+
+	switch {
```

Comment:
That's a nice addition! 
I discovered recentely that kubernetes already implements something like this: `cache.FilteringResourceEventHandler`.
It's implementation is really close to what you did: https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L264
WDYT of using this helper instead?